Reusable metatesting for add-on developers

mroth opened the pull request:
closes #80, this is the sample code for my proposal to make the `SuiteContext` itself general-purpose with an (optional!) variadic slice of additionalContextInitializers.

I believe this methodology will let a larger variety of developers make special-interest plugins to enhance the functionality of godog itself, and test them in a consistent fashion before offering to the public.

I personally would greatly prefer this general approach versus the of using external files and parsing jUnit XML proposed in https://github.com/DATA-DOG/godog/issues/80#issuecomment-299273733 for the following reasons:
1. No need to reinvent the wheel, potentially introducing bugs in the meta-testing framework itself. (As the existing SuiteContext is used for testing godog itself, it gives piece of mind to me as a add-on developer to know the same code is being used to test a potential godog add-on that it will work properly.
2. No need to write features to the filesystem enables easier concurrent-safe tests (due to filename collisions).
3. Having to develop a one-off test-harness to reliably replicate `SuiteContext` functionality externally feels discouraging to potential plugin developers, substantially increasing the scope of work needed to make a 
4. Easy re-use/discoverability by encouraging future add-on developers.

Cons:
1. End-user developers with no interest in developing godog addons may see `SuiteContext` in the godoc browser and it may increase the perceived complexity (I believe this could be potentially mitigated by renaming the function to something more obvious and writing documentation. I would be willing to help take this on if desired.)

That said, this is just my opinion. So thank you for your consideration! 🙇 

(This PR would at least need to be rebased off most-recent master before merging, but wanted to put it in a PR format to make it easier to review/discuss beforehand.)

l3pp4rd commented on suite_context.go:
```diff
@@ -0,0 +1,486 @@
+// provided for those developing add-on libraries for godog.
+//
+// For an example of how to use, see godog's own `features/` and `suite_test.go`.
+func SuiteContext(s *Suite, additionalContextInitializers ...func(suite *Suite)) {
```
Lets change the doc a little

``` go
// SuiteContext provides steps for godog suite execution and
// can be used for meta-testing of godog features/steps themselves.
//
// Beware, steps or their definitions might change without backward
// compatibility guarantees. A typical user of the godog library should never
// need this, rather it is provided for those developing add-on libraries for godog.
//
// For an example of how to use, see godog's own `features/` and `suite_test.go`.
func SuiteContext(s *Suite, additionalContextInitializers ...func(suite *Suite)) {
```

l3pp4rd commented on suite_context.go:
```diff
@@ -0,0 +1,486 @@
+	s.Step(`^the following events should be fired:$`, c.thereShouldBeEventsFired)
+
+	// lt
+	s.Step(`^savybių aplankas "([^"]*)"$`, c.featurePath)
```
maybe some of those steps aren't necessary, probably can be extracted later for godog specific test cases. but lets leave it for now just for the sake of simplistic pull request

l3pp4rd commented:
please rebase and update the godoc, then we can merge. I'll later try to extract some of the steps, which aren't needed for developers, like ones related to cucumber json formatter tests or duplicates for multilingual features.

mroth commented:
Rebased but having some weirdness with the line numbers in the `cucumber.feature`. Interesting the line number matching differs when run with `-race -covermode=atomic` versus standard `go test`!  I updated to match the line numbers with that, which makes it pass on 1.8 in travis-ci, but not on 1.5-1.7, which has differing line numbers as well!  @l3pp4rd any idea what's going on here?

(Maybe I need more :coffee:)

l3pp4rd commented on suite_context.go:
```diff
@@ -0,0 +1,488 @@
+}
+
+func (s *suiteContext) theRenderJSONWillBe(docstring *gherkin.DocString) error {
+	loc := regexp.MustCompile(`"suite_test.go:\d+"`)
```
update to `suite_context.go`

l3pp4rd replied:
when there is -race option, go adds more imports on the fly. and the line differs

mroth replied:
updated to `suite_context.go`, weirdly enough looks like there is a difference between line numbers on 1.5-1.7 and 1.8.  Perhaps 1.8 changed the imports required?

mroth commented:
Okay, was missing one regex.  Fixed now.  Going to squash all the line number fun in a second.

mroth commented:
Squashed! 🐛

l3pp4rd commented:
cool, thanks
